Regenerate hooks list documentations

tleon opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop developer documentation! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines on how to contribute:
https://devdocs.prestashop-project.org/8/contribute/documentation/how/
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 8.x
| Description?  | Regenerate hooks list documentations
| Fixed ticket? | Fixes #37163

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

This was generated using the command from the core with:
```sh
./bin/console prestashop:update:hooks-documentation /Users/myUser/devdocs-site/src/content/9/modules/concepts/hooks/list-of-hooks
```

nicosomb commented:
Why on 8.0.x branch?

jolelievre reviewed (changes requested):
I think the command has room for improvements:
- it is removing some valid documentation like `action<ClassName><Action>Before.md`, these hooks were detected but skipped in the command so I guess that's why they are removed here
- we didn't have infos extracted in the command bu here we have some data in the md files, I think the generation command should be able to also parse the data from the documentation from the md with regexp This way if we are not able to generate the description or usages they can be manually added in the doc Then on next generation the command can extract these data and it would avoid erasing some valid content

I block the PR for now until we improve the command but it's a really good start it just need some updates once the command has been improved

Also as @nicosomb said is the target here 8.X on purpose? does it mean you ran this command from the 8.1.x branch? it seems unlikely since the command has only been add on 9.0.x so far so maybe it's the wonrg base branch

kpodemski reviewed (changes requested):
I'm blocking it because there are too many issues at the moment.

One thing you probably weren't aware of:
You need to implement a feature in the generator to let us ignore some of the hooks; some from the docs have additional information, and some examples have been added that cannot be generated.

Hlavtox commented:
Same opinion as others - there is many useful information in the docs that have been contributed, it would get lost.

tleon commented:
So I targeted the right branch and the right files. 

Regarding the data loss : 

I ran the doc generation command directly on the current version of the md folder ( previously I deleted all md in there and regenerated all of them. I was not aware that some of the md may have been modified by hand.) 

I'm afraid that even while running the command like this, the files already there have been erased as  before. 

What I could do is : modify the command to ignore the files already present in the folder ( matching on the name of the files ). This would preserve the md modified by hand but those ones would miss some informations that I added in the command. ( I think ;) ) 

Also Like it was suggested I can ignore the hook containing a placeholder in the name I did it in the generation of the xml file I can just do the same thing here.

How does that sound ? 

@kpodemski @Hlavtox @nicosomb @jolelievre

jolelievre commented:
Hi @tleon 

ok good to know that the removal was manual and not caused by the command. I think we can think later about how we could remove automatically the obsolete hooks but let's not get carried away for now.

Regarding the override of existing hooks your suggestion is interesting but I'm thinking we could go one step further, the idea would be that if the file is already present the command only updates the fields that are missing or empty. Any already existing data/text in the MD files would be kept by default so only new data would be used for override.

To implement this we could add a `HookDocExtractor` service that can read and parse the data for each hook based on each MD file name. Then when you create the data to be exported the priority would be for each field something like `Doc > XML fixture > Auto generated data > Default fallback`


On the side note while you're on this topic some people on slack noticed that some hook doc pages are empty like this one https://devdocs.prestashop-project.org/9/modules/concepts/hooks/list-of-hooks/displaypdftemplate/ which matches the `displayPDF<Template>` hook, one theory is that there is a problem because of the charrets `<>` but it's worth looking into it, it may not be the (only) reason because this one is fine https://devdocs.prestashop-project.org/9/modules/concepts/hooks/list-of-hooks/actionclassnameactionbefore/

kpodemski commented:
> To implement this we could add a HookDocExtractor service that can read and parse the data for each hook based on each MD file name. Then when you create the data to be exported the priority would be for each field something like Doc > XML fixture > Auto generated data > Default fallback

This may require some changes in how we format information about a given hook, and it sounds a bit complicated, but if it's feasible, and that it could be trusted, then it would ease work on maintaining the list

jolelievre commented:
After discussion with @tleon we decided that for now we will do a simple modification in the command The command checks if the MD file already exists it simply skips it, so the command only adds new hooks

This way we are sure that we don't remove some manually modified files that are likely more up to date

If we want to improve this doc generation it will be done in another PR but we'll have to think about what we clearly want and if there's some real value to do it

jolelievre reviewed (approved):
Ok for me but I'd like to have @kpodemski opinion on this one

kpodemski commented on modules/concepts/hooks/list-of-hooks/actionAddressGridDefinitionModifier.md:
```diff
@@ -0,0 +1,27 @@
+hookTitle: 'Modify address grid definition'
+files:
+    -
+        url: 'https://github.com/PrestaShop/PrestaShop/blob/9.0.x/'
```
I think it could be interesting to have an origin of this hook, it fits the format of action<GridId>GridDefinitionModifier and we know where the hook is.

kpodemski replied:
and if it is not possible or hard to achieve, then maybe we don't need to list all of them but improve the search mechanism further according to this information:

```
Search tip: Some hooks are generated dynamically, so their names are documented in a generic way.

For example, actionAdminCustomersFormModifier is documented as action<AdminControllerClassName>FormModifier. A regex based search has been implemented, and generic hooks should be matched. However, when you see a controller name or action in the hook name and you can’t find it, try searching for a part of the hook name, like FormModifier.
```

from what I see most of the hooks that were generated are action hooks with the dynamic names

kpodemski commented on modules/concepts/hooks/list-of-hooks/actionAddressGridDefinitionModifier.md:
````diff
@@ -0,0 +1,27 @@
+
+## Call of the Hook in the origin file
+
+```php
+;
+```
````
This looks like a mistake. Again, it would be good to have a source for this hook here.

kpodemski commented on modules/concepts/hooks/list-of-hooks/actionAdministrationPageUploadQuotaSave.md:
```diff
@@ -0,0 +1,28 @@
+check_exceptions: false
+chain: false
+origin: core
+description: 'This hook allows to modify data of administration page upload quota options form after it was saved
+      '
```
Probably z `trim` missing

kpodemski commented on modules/concepts/hooks/list-of-hooks/fooHook.md:
```diff
@@ -0,0 +1,27 @@
+hookTitle: ''
+files:
+    -
+        url: 'https://github.com/PrestaShop/PrestaShop/blob/9.0.x/tests/Unit/Core/Hook/HookDispatcherTest.php'
```
`tests` directory should be excluded :)

kpodemski commented on modules/concepts/hooks/list-of-hooks/displayTopColumn.md:
```diff
@@ -0,0 +1,27 @@
+---
+Title: displayTopColumn
```
It's odd that there is no origin of this hook, it doesn't look like a dynamic one

jolelievre replied:
Seems like it was related to `homeslider` maybe an old module for 1.6

But the hook itself is not in the core anymore, this one is a remnant from the hook.xml file not a hook found by the extractor

kpodemski commented on modules/concepts/hooks/list-of-hooks/displayProductPageDrawer.md:
```diff
@@ -0,0 +1,27 @@
+---
+Title: displayProductPageDrawer
```
again, it seems there is an issue with getting a source of the display hooks

jolelievre replied:
SAme it's only in the hook.xml file

kpodemski commented on modules/concepts/hooks/list-of-hooks/displayProductListFunctionalButtons.md:
```diff
@@ -0,0 +1,27 @@
+---
+Title: displayProductListFunctionalButtons
```
I checked and it seems there is a hook registered but there is no use of it anywhere in the code 🤔 
Not sure what to do about it. Maybe of it is a display hook and there is no use, we could indicate it somehow 🤔  or this could help us clean the hooks in the db

cc @jolelievre

jolelievre replied:
Same only found in some XMLs and in example module's code I wonder if we should remove from the hook.xml all the hooks that are no longer found in the code base (except for dynamic ones)

Hlavtox replied:
@kpodemski Isn't this the hook to show buttons on product list hover that was removed “for performance reasons”?

kpodemski commented on modules/concepts/hooks/list-of-hooks/displayCarrierList.md:
```diff
@@ -0,0 +1,27 @@
+---
```
I noticed that this is the hook that is part of the deprecated hooks, here's the full list:

```
/**
     * List of all deprecated hooks.
     *
     * @var array
     */
    protected static $deprecated_hooks = [
        // Back office
        'backOfficeFooter' => ['from' => '1.7.0.0'],
        'displayBackOfficeFooter' => ['from' => '1.7.0.0'],

        // Shipping step
        'displayCarrierList' => ['from' => '1.7.0.0'],
        'extraCarrier' => ['from' => '1.7.0.0'],

        // Payment step
        'hookBackBeforePayment' => ['from' => '1.7.0.0'],
        'hookDisplayBeforePayment' => ['from' => '1.7.0.0'],
        'hookOverrideTOSDisplay' => ['from' => '1.7.0.0'],

        // Product page
        'displayProductTabContent' => ['from' => '1.7.0.0'],
        'displayProductTab' => ['from' => '1.7.0.0'],

        // Order page
        'displayAdminOrderRight' => ['from' => '1.7.7.0'],
        'displayAdminOrderLeft' => ['from' => '1.7.7.0'],
        'displayAdminOrderTabOrder' => ['from' => '1.7.7.0'],
        'displayAdminOrderTabShip' => ['from' => '1.7.7.0'],
        'displayAdminOrderContentOrder' => ['from' => '1.7.7.0'],
        'displayAdminOrderContentShip' => ['from' => '1.7.7.0'],

        // Controller
        'actionGetProductPropertiesAfter' => ['from' => '1.7.8.0'],
    ];
    ```
I think it would be good to skip them in the generator.

kpodemski commented on modules/concepts/hooks/list-of-hooks/displayBackOfficeFooter.md:
```diff
@@ -0,0 +1,27 @@
+check_exceptions: false
+chain: false
+origin: core
+description: 'This hook is displayed within the admin panel''s footer'
```
```suggestion
description: 'This hook is displayed within the admin panel''s footer'
```

I think this double `'` requires an escape

jolelievre replied:
Actually the double `''` is the way to escape the single `'`
Or am I missing something?

kpodemski reviewed (changes requested):
I added a few comments, there are some critical issues. I am also not sure if generating all the action hooks is a good idea... It makes the list huge, but if we could at least solve the problem with origin call and url to the file then it could help a lot.

jolelievre committed:
Auto generation of hooks documentation

kpodemski reviewed (approved):
Thank you guys @tleon @jolelievre 

I don't have any comments apart from this double `;;`, do you think you can regenerate the list @jolelievre without it?

jolelievre commented:
@kpodemski fixed on the core PR https://github.com/PrestaShop/PrestaShop/pull/38232 and I updated this one as well, let me know if it's all good so we can finally close this topic 😅

kpodemski commented:
merged! thank you @jolelievre @tleon ❤️
